### PR TITLE
Feature/login

### PR DESCRIPTION
**Pull request by dev-afk-now**

### CODE REVIEW

File: Blanball/AppDelegate.swift [hunk 1]
```diff
@@ -6,18 +6,25 @@
...
-        
+    func application(
+        _ application: UIApplication,
+        didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
+    ) -> Bool {
         let window = UIWindow(frame: UIScreen.main.bounds)
         self.window = window
         appCoordinator = AppCoordinator()
         appCoordinator.setRoot(for: window)
+//        self.window?.rootViewController = LoadingViewController(nibName: "LoadingViewController", bundle: nil)
```
Comment (GodLizarx) [inline Blanball/AppDelegate.swift]: remove old code

File: Blanball/Modules/Loading/LoadingViewController.swift [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    // MARK: - Outlets -
+    
+    @IBOutlet private weak var titleLabel: UILabel!
+    @IBOutlet private weak var spinnerImageView: UIImageView!
+    
+    // MARK: - Lifecycle -
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        NotificationCenter.default.addObserver(
```
Comment (GodLizarx) [inline Blanball/Modules/Loading/LoadingViewController.swift]: should be stored in "token: NSObjectProtocol" and invalidated in deinit

File: Blanball/Modules/Login/ViewController/LoginViewController.swift [hunk 3]
```diff
@@ -10,21 +10,99 @@ import Combine
...
         super.viewDidLoad()
         viewModel.start()
+        configureUI()
     }
+
+    override func setupView() {}
     
-    override func setupView() {
-        
+    private func configureUI() {
```
Comment (GodLizarx) [inline Blanball/Modules/Login/ViewController/LoginViewController.swift]: should use setupView method

File: Blanball/Modules/Login/ViewModel/LoginViewModel.swift [hunk 4]
```diff
@@ -7,10 +7,11 @@
 
 import Foundation
 import XCoordinator
+import Combine
 
 final class LoginViewModel: BaseViewModel<LoginViewModelState> {
     
-    private let router: UnownedRouter<LoginRoute>
+    private let router: UnownedRouter<LoginRoute>!
```
Comment (GodLizarx) [inline Blanball/Modules/Login/ViewModel/LoginViewModel.swift]: replace with StrongRouter<LoginRoute>?

File: Blanball/Views/BlanRoundedTextField/BlanRoundedTextField+Constants.swift [hunk 5]
```diff
@@ -0,0 +1,41 @@
+//
+//  BlanRoundedTextField+Constants.swift
+//  Blanball
+//
+//  Created by Nik Dub on 10.03.2023.
+//
+
+import UIKit
+
+extension BlanRoundedTextField {
```
Comment (GodLizarx) [inline Blanball/Views/BlanRoundedTextField/BlanRoundedTextField+Constants.swift]: rename with TitledTextField

File: Blanball/Views/BlanRoundedTextField/BlanRoundedTextField.swift [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+        titleLabel.textColor = isValid
+        ? Constants.titleColor
+        : Constants.errorColor
+        self.errorLabelHeightConstraint.constant = isValid ? .zero : 20
+        UIView.animate(withDuration: 0.1, delay: .zero, options: .curveEaseInOut) {
+            self.view.layoutIfNeeded()
+        }
+    }
+    
+    // TODO: Modify regex characters
```
Comment (GodLizarx) [inline Blanball/Views/BlanRoundedTextField/BlanRoundedTextField.swift]: remove TODO

File: Blanball/Views/BlanRoundedTextField/BlanRoundedTextField.swift [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        let test = NSPredicate(format: "SELF MATCHES %@", literalRegex)
+        return test.evaluate(with: string)
+    }
+    
+    private func setupConstraints() {
+        addSubview(editingTextField)
+        addSubview(labelContainerView)
+        addSubview(errorLabel)
+        labelContainerView.addSubview(titleLabel)
+        NSLayoutConstraint.activate([
```
Comment (GodLizarx) [inline Blanball/Views/BlanRoundedTextField/BlanRoundedTextField.swift]: SnapKit

File: Blanball/Views/BlanRoundedTextField/BlanRoundedTextFieldState.swift [hunk 8]
```diff
@@ -0,0 +1,13 @@
+//
+//  BlanRoundedTextFieldState.swift
+//  Blanball
+//
+//  Created by Nik Dub on 06.03.2023.
+//
+
+import Foundation
+
+enum BlanRoundedTextFieldState {
```
Comment (GodLizarx) [inline Blanball/Views/BlanRoundedTextField/BlanRoundedTextFieldState.swift]: rename

File: Blanball/Views/CommonButton/CommonButton.swift [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    
+    override var isEnabled: Bool {
+        didSet {
+            backgroundColor = isEnabled
+            ? switchedBackgroundColor.0
+            : switchedBackgroundColor.1
+        }
+    }
+    
+    private var switchedBackgroundColor: (UIColor?, UIColor?) = (nil, nil) {
```
Comment (GodLizarx) [inline Blanball/Views/CommonButton/CommonButton.swift]: example of naming: backgroundColors: (onColor: UIColor?, offColor: UIColor?) = (nil, nil)

File: Blanball/Views/CommonLabel/CommonLabel.swift [hunk 10]
```diff
@@ -0,0 +1,20 @@
+//
+//  CommonLabel.swift
+//  Blanball
+//
+//  Created by Nik Dub on 10.03.2023.
+//
+
+import UIKit
+
+class CommonLabel: UILabel {
```
Comment (GodLizarx) [inline Blanball/Views/CommonLabel/CommonLabel.swift]: use Extension to UILabel if it doesn't change behaviour

File: Blanball/Views/LabeledCheckBox/LabeledCheckBox.swift [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+import UIKit
+
+class LabeledCheckBox: UIView {
+    
+    // MARK: - Internal properties -
+    
+    @Published private(set) var state: LabeledCheckBoxState
+    
+    // MARK: - Private properties -
```
Comment (GodLizarx) [inline Blanball/Views/LabeledCheckBox/LabeledCheckBox.swift]: checkbox should be separate from label, just add ATapGesture to their container for single action

File: Blanball/Views/LabeledCheckBox/LabeledCheckBox.swift [hunk 12]
```diff
@@ -0,0 +1,110 @@
+//
+//  LabeledCheckBox.swift
+//  Blanball
+//
+//  Created by Nik Dub on 10.03.2023.
+//
+
+import UIKit
+
+class LabeledCheckBox: UIView {
```
Comment (GodLizarx) [inline Blanball/Views/LabeledCheckBox/LabeledCheckBox.swift]: rename just CheckBox

File: Blanball/Views/UnderlinedDashButton/UnderlinedDashButton.swift [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+            ),
+            for: .normal
+        )
+    }
+    
+    // MARK: - Private properties -
+    
+    private func setupConstraints() {
+        addSubview(mainButton)
+        NSLayoutConstraint.activate([
```
Comment (GodLizarx) [inline Blanball/Views/UnderlinedDashButton/UnderlinedDashButton.swift:62]: SnapKit

**Review by GodLizarx — CHANGES_REQUESTED** [review]

File: Blanball/AppDelegate.swift (hunk 1 above)
Comment (dev-afk-now) [reply Blanball/AppDelegate.swift]: Fixed in `Changes According PR#7`

File: Blanball/Modules/Loading/LoadingViewController.swift (hunk 2 above)
Comment (dev-afk-now) [reply Blanball/Modules/Loading/LoadingViewController.swift]: Transitioned NotificationCenter call to separate func

File: Blanball/Modules/Login/ViewController/LoginViewController.swift (hunk 3 above)
Comment (dev-afk-now) [reply Blanball/Modules/Login/ViewController/LoginViewController.swift]: Fixed in `Changes According PR#7`

File: Blanball/Modules/Login/ViewModel/LoginViewModel.swift (hunk 4 above)
Comment (dev-afk-now) [reply Blanball/Modules/Login/ViewModel/LoginViewModel.swift]: Fixed in Changes `According PR#7`

File: Blanball/Views/BlanRoundedTextField/BlanRoundedTextField+Constants.swift (hunk 5 above)
Comment (dev-afk-now) [reply Blanball/Views/BlanRoundedTextField/BlanRoundedTextField+Constants.swift]: Fixed in Changes `According PR#7`

File: Blanball/Views/BlanRoundedTextField/BlanRoundedTextField.swift (hunk 6 above)
Comment (dev-afk-now) [reply Blanball/Views/BlanRoundedTextField/BlanRoundedTextField.swift]: Fixed in `Changes According PR#7`

File: Blanball/Views/BlanRoundedTextField/BlanRoundedTextFieldState.swift (hunk 8 above)
Comment (dev-afk-now) [reply Blanball/Views/BlanRoundedTextField/BlanRoundedTextFieldState.swift]: Fixed in `Changes According PR#7`

File: Blanball/Views/CommonButton/CommonButton.swift (hunk 9 above)
Comment (dev-afk-now) [reply Blanball/Views/CommonButton/CommonButton.swift]: New names:

> backgroundColors, borderColors,  tintColors

`Changes According PR#7`

File: Blanball/Views/CommonLabel/CommonLabel.swift (hunk 10 above)
Comment (dev-afk-now) [reply Blanball/Views/CommonLabel/CommonLabel.swift]: Transitioned into extension for UILabel in `Changes According PR#7`

File: Blanball/Views/LabeledCheckBox/LabeledCheckBox.swift (hunk 11 above)
Comment (dev-afk-now) [reply Blanball/Views/LabeledCheckBox/LabeledCheckBox.swift]: Fixed in `Changes According PR#7`

File: Blanball/Views/LabeledCheckBox/LabeledCheckBox.swift (hunk 12 above)
Comment (dev-afk-now) [reply Blanball/Views/LabeledCheckBox/LabeledCheckBox.swift]: Fixed in `Changes According PR#7`

**Pull request merged**